### PR TITLE
*: fix issue of stale peer block resolve-ts cause by ignore gc message

### LINKED ISSUE #16504: Stale peer may cause resolve ts doesn't push
### What version of TiKV are you using?

v6.5.4

### Steps to reproduce

image

Related log are following:

```log
"[2024/01/26 07:12:21.299 +00:00] [INFO] [store.rs:624] [\"raft message is stale, tell to gc\"] [to_peer_store_id=277] [to_peer_id=1041669428] [msg_type=MsgAppendResponse] [current_region_epoch=\"conf_ver: 208216 version: 6072877\"] [region_id=1041669023]"
```

```log
"[2024/01/26 07:12:21.280 +00:00] [INFO] [apply.rs:2385] [\"conf change successfully\"] [\"current region\"=\"id: 1041669023 start_key: ? end_key: ? region_epoch { conf_ver: 208211 version: 6072877 } peers { id: 1041669428 store_id: 277 role: DemotingVoter } peers { id: 1041669430 store_id: 477403250 role: DemotingVoter } peers { id: 1041669429 store_id: 478557948 } peers { id: 1041672161 store_id: 282 role: IncomingVoter } peers { id: 1041672162 store_id: 20296629 role: IncomingVoter }\"] [\"original region\"=\"id: 1041669023 start_key: ? end_key: ? region_epoch { conf_ver: 208207 version: 6072877 } peers { id: 1041669428 store_id: 277 } peers { id: 1041669430 store_id: 477403250 } peers { id: 1041669429 store_id: 478557948 } peers { id: 1041672161 store_id: 282 role: Learner } peers { id: 1041672162 store_id: 20296629 role: Learner }\"] [changes=\"[peer { id: 1041672161 store_id: 282 }, peer { id: 1041672162 store_id: 20296629 }, change_type: AddLearnerNode peer { id: 1041669428 store_id: 277 role: Learner }, change_type: AddLearnerNode peer { id: 1041669430 store_id: 477403250 role: Learner }]\"] [peer_id=1041669428] [region_id=1041669023]" 

"[2024/01/26 07:12:21.287 +00:00] [INFO] [apply.rs:2415] [\"leave joint state successfully\"] [region=\"id: 1041669023 start_key: ? end_key: ? region_epoch { conf_ver: 208215 version: 6072877 } peers { id: 1041669428 store_id: 277 role: Learner } peers { id: 1041669430 store_id: 477403250 role: Learner } peers { id: 1041669429 store_id: 478557948 } peers { id: 1041672161 store_id: 282 } peers { id: 1041672162 store_id: 20296629 }\"] [peer_id=1041669428] [region_id=1041669023]" 

"[2024/01/26 07:12:21.300 +00:00] [INFO] [peer.rs:2951] [\"receive stale gc message, ignore.\"] [peer_id=1041669428] [region_id=1041669023]"
...
...
"[2024/01/26 07:12:35.591 +00:00] [INFO] [endpoint.rs:888] [\"the max gap of safe-ts is large\"] [\"min start ts\"=None] [\"lock num\"=None] [advance-ts-interval=ReadableDuration(2s)] [\"region id\"=1041669023] [\"oldest safe-ts\"=447284022796091414] [gap=16851]"
...
...
"[2024/01/26 07:31:46.352 +00:00] [INFO] [endpoint.rs:888] [\"the max gap of safe-ts is large\"] [\"min start ts\"=None] [\"lock num\"=None] [advance-ts-interval=ReadableDuration(2s)] [\"region id\"=1041669023] [\"oldest safe-ts\"=447284022796091414] [gap=1167601]"
...
...

"[2024/01/26 07:31:47.457 +00:00] [INFO] [store.rs:624] [\"raft message is stale, tell to gc\"] [to_peer_store_id=277] [to_peer_id=1041669428] [msg_type=MsgHup] [current_region_epoch=\"conf_ver: 208217 version: 6072877\"] [region_id=1041669023]"
"[2024/01/26 07:31:47.457 +00:00] [INFO] [peer.rs:2960] [\"receives gc message, trying to remove\"] [to_peer=\"id: 1041669428 store_id: 277 role: Learner\"] [peer_id=1041669428] [region_id=1041669023]"
"[2024/01/26 07:31:47.457 +00:00] [INFO] [apply.rs:3723] [\"remove delegate from apply delegates\"] [peer_id=1041669428] [region_id=1041669023]"
"[2024/01/26 07:31:47.457 +00:00] [INFO] [peer.rs:1461] [\"begin to destroy\"] [peer_id=1041669428] [region_id=1041669023]"
"[2024/01/26 07:31:47.460 +00:00] [INFO] [peer.rs:1565] [\"peer destroy itself\"] [keep_data=false] [clean=true] [takes=2.868957ms] [peer_id=1041669428] [region_id=1041669023]"
"[2024/01/26 07:31:47.460 +00:00] [INFO] [router.rs:345] [\"[region 1041669023] shutdown mailbox\"]"
```

The region peers that safe-ts did not advance were 1041669428.  And the peer 1041669428 received a gc message at `07:12:21.300`, but ignored it, and then the peer's safe-ts started not advancing(blocked resolve ts too). Then the peer sent itself a GC message until `07:31:47.457` and started destroying the peer.

Why does the peer ignore gc message when it receives it at the first time  `07:12:21.300`? I think the [check code](https://github.com/tikv/tikv/blob/0812e2f85633093991c01b826c77392aff7bfc02/components/raftstore/src/store/fsm/peer.rs#L2944) has some bug. 

```rust
        if self.fsm.peer.peer != *msg.get_to_peer() {
            info!(
                "receive stale gc message, ignore.";
                "region_id" => self.fsm.region_id(),
                "peer_id" => self.fsm.peer_id(),
            );
            self.ctx.raft_metrics.message_dropped.stale_msg.inc();
            return;
        }
```

The check should only compare the peer id instead of comparing all peer info, like this PR change https://github.com/tikv/tikv/pull/16505? I see that the role of this peer changes in the log. At `07:12:21.280`, the role becomes `DemotingVoter`, and at `07:12:21.287`, the role becomes `Learner`. Since the role changed, with upper check, the stale peer will ignore the gc message and start block resolve ts to push.

### PR DESCRIPTION
**Pull request by crazycs520**
### What is changed and how it works?

Issue Number: Close #16504

What's Changed:

```commit-message
Fix issue of stale peer block resolve-ts cause by ignore gc message.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [x] Need to cherry-pick to the release branch

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
Fix issue of stale peer block resolve-ts caused by ignore gc message.
```

### CODE REVIEW

File: components/raftstore/src/store/fsm/peer.rs [hunk 1]
```diff
@@ -3164,11 +3164,12 @@ where
             return;
         }
 
-        if self.fsm.peer.peer != *msg.get_to_peer() {
+        if self.fsm.peer.peer.get_id() != msg.get_to_peer().get_id() {
```
Comment (cfzjywxk) [inline components/raftstore/src/store/fsm/peer.rs:3167]: Need to cover the case in tests.

Comment (Connor1996) [reply components/raftstore/src/store/fsm/peer.rs:3167]: +1

Comment (crazycs520) [reply components/raftstore/src/store/fsm/peer.rs:3167]: Sorry, I tried to write a test, but I don't know how to construct the scenario in issue #16504.

Comment (cfzjywxk) [reply components/raftstore/src/store/fsm/peer.rs:3167]: @crazycs520 
I could add one when I have time.

**Comment by nkg-** [comment]
Any updates on this.

File: components/raftstore/src/store/fsm/peer.rs (hunk 1 above)
Comment (cfzjywxk) [reply components/raftstore/src/store/fsm/peer.rs:3178]: @Connor1996 
An integration test case is added, PTAL

**Comment by cfzjywxk** [comment]
@Connor1996 @overvenus @tonyxuqqi 
 PTAL

**Comment by BornChanger** [comment]
> @BornChanger : This PR should reduce blocking of resolve-ts right? Do we need it to reduce the EBS BR backup failures due to "GC safepoint error"
Yes, I think this PR helps.

**Review by SpadeA-Tang — APPROVED** [review]

**Review by Connor1996 — APPROVED** [review]
LGTM

**Comment by nkg-** [comment]
Can we cherrypick to release6.5 branch please.

**Comment by nkg-** [comment]
FYI This PR depends on https://github.com/tikv/tikv/pull/15057, specifically wait_peer_state function, added in the unit test. And that PR is not in release6.5.

**Comment by nkg-** [comment]
Can we already cherrypick this to release-7.5. cc: @BornChanger

**Pull request merged**